core/src/main/java/hudson/util/AtomicFileWriter.java

```diff
@@ -108,35 +119,72 @@ public void close() throws IOException {
      * the {@link #commit()} is called, to simplify coding.
      */
     public void abort() throws IOException {
-        close();
-        tmpFile.delete();
+        closeAndDeleteTempFile();
     }
 
     public void commit() throws IOException {
         close();
-        if (destFile.exists()) {
+        try {
+            // Try to make an atomic move.
+            Files.move(tmpPath, destPath, StandardCopyOption.ATOMIC_MOVE);
+        } catch (IOException e) {
+            // If it falls here that can mean many things. Either that the atomic move is not supported,
+            // or something wrong happened. Anyway, let's try to be over-diagnosing
+            LOGGER.log(Level.WARNING, "Unable to move atomically, falling back to non-atomic move.");
```

Comment:
I had a suspicion that this would cause a lot of logging for people that use certain (CIFS) network shares as Jenkins home
but tested writing to a cifs share and the ATOMIC_MOVE does not report an error.

```
        File source = new File("\\\\nas\\home\\test\\one.txt");
        File target = new File("\\\\nas\\home\\test\\two.txt");
        final Path sourcePath = source.toPath();
        final Path targetPath = target.toPath();
        
        Files.write(sourcePath, "source".getBytes());
        Files.write(targetPath, "target".getBytes());

        assertTrue(source.exists());
        assertTrue(target.exists());

        Files.move(sourcePath, targetPath, StandardCopyOption.ATOMIC_MOVE);

```